plugins/tpm-eventlog/fu-tpm-eventlog-device.c

```diff
@@ -0,0 +1,355 @@
+	if (blobstr != NULL)
+		fu_common_string_append_kv (str, idt, "BlobStr", blobstr);
+}
+
+static void
+fu_tpm_eventlog_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuTpmEventlogDevice *self = FU_TPM_EVENTLOG_DEVICE (device);
+	if (self->items->len > 0) {
+		fu_common_string_append_kv (str, idt, "Items", NULL);
+		for (guint i = 0; i < self->items->len; i++) {
+			FuTpmEventlogDeviceItem *item = g_ptr_array_index (self->items, i);
+			fu_tpm_eventlog_device_item_to_string (item, idt + 1, str);
+		}
+	}
+}
+
+gchar *
+fu_tpm_eventlog_device_report_metadata (FuTpmEventlogDevice *self)
+{
+	GString *str = g_string_new ("");
+	for (guint i = 0; i < self->items->len; i++) {
+		FuTpmEventlogDeviceItem *item = g_ptr_array_index (self->items, i);
+		g_autofree gchar *blobstr = fu_tpm_eventlog_device_blobstr (item->blob);
+		g_string_append_printf (str, "0x%08x %s", item->kind, item->checksum);
+		if (blobstr != NULL)
+			g_string_append_printf (str, " [%s]", blobstr);
+		g_string_append (str, "\n");
+	}
+	if (str->len > 0)
+		g_string_truncate (str, str->len - 1);
+	return g_string_free (str, FALSE);
+}
+
+static void
+fu_tpm_eventlog_device_item_free (FuTpmEventlogDeviceItem *item)
+{
+	g_bytes_unref (item->blob);
+	g_free (item->checksum);
+	g_free (item);
+}
+
+static void
+fu_tpm_eventlog_device_init (FuTpmEventlogDevice *self)
+{
+	self->items = g_ptr_array_new_with_free_func ((GDestroyNotify) fu_tpm_eventlog_device_item_free);
+	fu_device_set_name (FU_DEVICE (self), "Event Log");
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_set_physical_id (FU_DEVICE (self), "DEVNAME=/dev/tpm0");
+	fu_device_set_logical_id (FU_DEVICE (self), "eventlog");
+	fu_device_add_parent_guid (FU_DEVICE (self), "system-tpm");
+	fu_device_add_instance_id (FU_DEVICE (self), "system-tpm-eventlog");
+}
+
+static void
+fu_tpm_eventlog_device_finalize (GObject *object)
+{
+	FuTpmEventlogDevice *self = FU_TPM_EVENTLOG_DEVICE (object);
+
+	g_ptr_array_unref (self->items);
+
+	G_OBJECT_CLASS (fu_tpm_eventlog_device_parent_class)->finalize (object);
+}
+
+static void
+fu_tpm_eventlog_device_class_init (FuTpmEventlogDeviceClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	FuDeviceClass *klass_device = FU_DEVICE_CLASS (klass);
+	object_class->finalize = fu_tpm_eventlog_device_finalize;
+	klass_device->to_string = fu_tpm_eventlog_device_to_string;
+}
+
+static gchar *
+fu_common_read_strhex_safe (const guint8 *buf,
+			    gsize bufsz,
+			    gsize offset,
+			    gsize length,
+			    GError **error)
+{
+	g_autoptr(GString) csum = g_string_new (NULL);
+	g_autofree guint8 *digest = g_malloc0 (length);
+	if (!fu_memcpy_safe (digest, length, 0x0,	/* dst */
+			     buf, bufsz, offset,	/* src */
+			     length, error))
+		return FALSE;
+	for (guint i = 0; i < length; i++)
+		g_string_append_printf (csum, "%02x", digest[i]);
+	return g_string_free (g_steal_pointer (&csum), FALSE);
+}
```

Comment:
I guess you meant for this function to go into fu-common, but forgot to move it?

Reply:
I wasn't sure. I kinda wanted to leave it there, and only export it if another plugin needed the same functionality.

Reply:
OK, I'd say probably rename it to something without `fu_common` prefix then